doc/CodeReview.md

```diff
@@ -172,7 +182,190 @@ A meta target that runs every piece of CI possible.
+members and other interested parties. Likely, only trivial pull requests (e.g.
+spelling/formatting fixes, test fixes, or documentation/comment updates) will be
+merged without discussion.
+
+The types and detail of the comments will vary based on the type of change and
+the location within the repository of the changes. Changes marked "significant"
+will receive a more thorough review. Similarly, changes to code that is used
+widely (e.g. changes within the core `kernel` crate) will be more scrutinized.
+
+To help pull request reviewers and pull request submitters alike, we document
+review principles that will be used when evaluating pull requests.
+
+### General Review Principles
+
+**PR Mechanics**
+
+- Is this a self-contained change, or should portions of the pull request be
+  split into separate PRs? Note, this is not evaluated by number of files or
+  changed lines, but rather by the semantic meaning of the changes.
+- Are the commits all relevant to the change, or are there possibly unrelated
+  branches that are unintentionally included?
+- Does the provide enough explanation to help reviewers understand its purpose,
+  and to explain the change for future readers of the code? Does the PR link to
+  relevant tracking issues or other discussions? Are there existing discussions
+  that should be referenced?
+
+**Documentation and Comments**
+
+- Many core designs of Tock are document in specific markdown files. Does this
+  PR change any of those designs/details, and are the corresponding documents
+  updated?
+- Does the change include proper rustdoc comments for new files and new data
+  structures?
+- If the change is user-facing (i.e. part of a public API or command line
+  interface) does it include enough helpful information for users to understand
+  how to use it?
+- If documentation is removed, is it clear that it is no longer accurate or
+  needed? Or is there justification for removing the documentation?
+
+**Code**
+
+- Is `unsafe` used? If so, the changes in this PR need to be carefully
+  evaluated. The purpose of `unsafe` must be documented and why it is used
+  correctly must be explained in a comment.
+- Is `unsafe` used when code is not actually memory or type unsafe (i.e. does
+  not violate the Rust safety model)? This should not be marked `unsafe` and
+  instead should likely be marked safe or use a capability.
+- Does the code use interrupts and callbacks? If so, the code MUST NOT issue a
+  callback from a downcall. The callback may ONLY be called in response to an
+  interrupt. Using deferred calls is often necessary to remedy this.
+- Are Rust features (i.e. conditional compilation and `#[cfg]`) used? These must
+  be clearly motivated and documented, and are only permitted in specific cases.
+- Are `lib.rs` or `mod.rs` files added? In general these should only be used to
+  reference other modules and setup exports. Actual OS logic should be in
+  descriptively named files.
```

Comment:
I know that we've been generally following this rule, but Rust's module system sometimes makes it very awkward to require `mod.rs` at `some::path::in::the::module::hierarchy`, but then require an extra `::spurious_module_path` at the end, just because we add another descriptively named file. In many ways we can work around this through reexports, but we also have counterexamples in the repository.

The current text says "in general these should", and that seems like fine language to me, if it's indeed not actually meant to be a hard rule.

Reply:
We have a legacy dating back to the beginnings of tock of using lib.rs in arch crates which I guess is ok, hence my hedging. (Although, actually looking at the commits my excuse isn't really valid, we _had_ a different file and moved it to lib.rs :( https://github.com/tock/tock/commit/2c1d6206d91f57a5d402502617e42b0222ffcdc0)

That being said, for me it is a rule: either you have little code/modularity and do not need a crate/folder, or you have enough code/modularity to warrant a crate/folder and the structure should plan for mode code to be added in the future (i.e. do not use the mod/lib file).